### PR TITLE
8292575: riscv: Represent Registers as values

### PR DESCRIPTION
**Pull request by feilongjiang**
Follow up [JDK-8292153](https://bugs.openjdk.org/browse/JDK-8292153) and [JDK-8292203](https://bugs.openjdk.org/browse/JDK-8292203), migrate away from pointer-based representation of Register values.

It improves compile-time checking by forbidding implicit conversions between integrals and pointers.

Testing:
- hotspot/jdk tier1 on unmatched board
- hotspot/jdk tier1 on qemu

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292575](https://bugs.openjdk.org/browse/JDK-8292575): riscv: Represent Registers as values

### Reviewers
 * [Yanhong Zhu](https://openjdk.org/census#yzhu) (@yhzhu20 - Author) ⚠️ Review applies to [b9dbd6b4](https://git.openjdk.org/jdk/pull/9936/files/b9dbd6b47ac3cb266ba445125532acd375473109)
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9936/head:pull/9936` \
`$ git checkout pull/9936`

Update a local copy of the PR: \
`$ git checkout pull/9936` \
`$ git pull https://git.openjdk.org/jdk pull/9936/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9936`

View PR using the GUI difftool: \
`$ git pr show -t 9936`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9936.diff">https://git.openjdk.org/jdk/pull/9936.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/vmreg_riscv.inline.hpp [hunk 1]
```diff
@@ -26,20 +26,17 @@
...
-    return VMRegImpl::Bad();
-  }
-  return VMRegImpl::as_VMReg(encoding() * RegisterImpl::max_slots_per_register);
+inline VMReg Register::RegisterImpl::as_VMReg() const {
+  return VMRegImpl::as_VMReg(encoding() * Register::max_slots_per_register);
 }
 
-inline VMReg FloatRegisterImpl::as_VMReg() const {
-  return VMRegImpl::as_VMReg((encoding() * FloatRegisterImpl::max_slots_per_register) +
+inline VMReg FloatRegister::FloatRegister::FloatRegisterImpl::as_VMReg() const {
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/vmreg_riscv.inline.hpp:33]: Shouldn't this be: inline VMReg FloatRegister::FloatRegisterImpl::as_VMReg() const?

Comment (feilongjiang) [reply src/hotspot/cpu/riscv/vmreg_riscv.inline.hpp]: Yes, fixed.

**Review by yhzhu20 — APPROVED** [review]
Looks good.

**Comment by yadongw** [comment]
lgtm(not a reviewer)

**Review by yadongw — APPROVED** [review]

**Review by RealFYang — APPROVED** [review]
Updated changes look good. Thanks.

**Pull request closed without merging**